### PR TITLE
Implement more flows ops methods part 3

### LINKED ISSUE #68: implement more flows ops methods - part 3
implement methods:
* groupedWithin
* groupedWeightedWithin
* groupBy

### PR DESCRIPTION
**Pull request by emil-bar**
Closes #68

### CODE REVIEW

File: flows/src/main/java/com/softwaremill/jox/flows/Flow.java [hunk 1]
```diff
@@ -268,6 +271,138 @@ public <U> Flow<U> map(Function<T, U> mappingFunction) {
...
+     * @param <U> type of output flow
+     */
+    @FunctionalInterface
+    public interface StatefulMapper<T, S, U> {
+        /**
+         * @param state current state
+         * @param element current input flow element
+         * @return pair of new state and `element` mapped to new type `U`
+         */
+        Map.Entry<S, U> apply(S state, T element);
```
[the remark this answers is not in the document]
Comment (adamw) [reply flows/src/main/java/com/softwaremill/jox/flows/Flow.java:289]: Map.Entry sounds good - do we really use `null` anywhere?

[the remark this answers is not in the document]
Comment (emil-bar) [reply flows/src/main/java/com/softwaremill/jox/flows/Flow.java:289]: In ox we used null as marker for `lines` method. 
I used Optional for workaround https://github.com/softwaremill/jox/pull/77/files#diff-46d2537d747d24775cf70630d584ac42326ea6902d3ddd10f5a861841c2c3b6cR722

Overall we can leave it as it is

[the remark this answers is not in the document]
Comment (adamw) [reply flows/src/main/java/com/softwaremill/jox/flows/Flow.java:289]: Ah, Optional is better :) 👍

File: flows/src/main/java/com/softwaremill/jox/flows/Flow.java [hunk 2]
```diff
@@ -268,6 +271,138 @@ public <U> Flow<U> map(Function<T, U> mappingFunction) {
...
+
+    /**
+     * Functional interface used for {@link Flow#mapStateful} and {@link Flow#mapStatefulConcat}.
+     *
+     * @param <T> type of input flow elements
+     * @param <S> type of state
+     * @param <U> type of output flow
+     */
+    @FunctionalInterface
+    public interface StatefulMapper<T, S, U> {
```
[the remark this answers is not in the document]
Comment (adamw) [reply flows/src/main/java/com/softwaremill/jox/flows/Flow.java:283]: No, this looks good :)

**Pull request merged**